src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+        }
+    }
+
+    unsafe {
+        let mut g = Guard { len: buf.len(), s: buf.as_mut_vec() };
+        let ret = f(g.s);
+        if str::from_utf8(&g.s[g.len..]).is_err() {
+            ret.and_then(|()| {
+                Err(Error::new(ErrorKind::InvalidInput,
+                               "stream did not contain valid UTF-8", None))
+            })
+        } else {
+            g.len = g.s.len();
+            ret
+        }
+    }
+}
+
+fn read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> Result<()> {
+    loop {
+        if buf.capacity() == buf.len() {
+            buf.reserve(DEFAULT_BUF_SIZE);
+        }
+        match with_end_to_cap(buf, |b| r.read(b)) {
+            Ok(0) => return Ok(()),
+            Ok(_) => {}
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+            Err(e) => return Err(e),
+        }
+    }
+}
+
+/// A trait for objects which are byte-oriented sources.
+///
+/// Readers are defined by one method, `read`. Each call to `read` will attempt
+/// to pull bytes from this source into a provided buffer.
+///
+/// Readers are intended to be composable with one another. Many objects
+/// throughout the I/O and related libraries take and provide types which
+/// implement the `Read` trait.
+pub trait Read {
+    /// Pull some bytes from this source into the specified buffer, returning
+    /// how many bytes were read.
+    ///
+    /// This function does not provide any guarantees about whether it blocks
+    /// waiting for data, but if an object needs to block for a read but cannot
+    /// it will typically signal this via an `Err` return value.
+    ///
+    /// If the return value of this method is `Ok(n)`, then it must be
+    /// guaranteed that `0 <= n <= buf.len()`. If `n` is `0` then it indicates
+    /// that this reader as reached "end of file" and will likely no longer be
+    /// able to produce more bytes, but this is not guaranteed, however. A
+    /// nonzero `n` value indicates that the buffer `buf` has been filled in
+    /// with `n` bytes of data from this source.
+    ///
+    /// No guarantees are provided about the contents of `buf` when this
+    /// function is called, implementations cannot rely on any property of the
+    /// contents of `buf` being true.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. If an error is returned then it is guaranteed
+    /// that no bytes were read successfully.
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+
+    /// Read all remaining bytes in this source, placing them into `buf`.
+    ///
+    /// This function will continuously invoke `read` until `Ok(0)` or an error
+    /// is reached, at which point this function will immediately return.
+    ///
+    /// This function will only successfully return if an invocation of `Ok(0)`
+    /// succeeds at which point all bytes have been read into `buf`.
```

Comment:
>an invocation of `Ok(0)`

Maybe my English is failing me but this doesn't make much sense.

>at which point all bytes have been read into `buf`

All bytes of what?